firewood/src/storage/linear/mod.rs

```diff
@@ -68,47 +70,60 @@ mod filebacked;
 mod historical;
 mod layered;
 mod proposed;
-#[cfg(test)]
-mod tests;
 
 #[derive(Debug)]
-pub(super) struct LinearStore<S: ReadLinearStore> {
-    state: S,
-}
-
-/// All linearstores support reads
-pub(super) trait ReadLinearStore: Send + Sync + Debug {
-    fn stream_from(&self, addr: u64) -> Result<Box<dyn Read + '_>, Error>;
-    fn size(&self) -> Result<u64, Error>;
+pub(super) enum ImmutableLinearStore {
+    Historical(Historical),
+    Proposed(Arc<Proposed<Immutable>>),
+    FileBacked(FileBacked),
+    Invalid, // TODO do we need this variant?
 }
 
-impl ReadLinearStore for Arc<dyn ReadLinearStore> {
-    fn stream_from(&self, addr: u64) -> Result<Box<dyn Read + '_>, Error> {
-        self.deref().stream_from(addr)
+impl ImmutableLinearStore {
+    pub fn stream_from(&self, addr: u64) -> Result<Box<dyn Read + '_>, Error> {
+        match self {
+            ImmutableLinearStore::Historical(historical) => historical.stream_from(addr),
+            ImmutableLinearStore::Proposed(proposed) => proposed.stream_from(addr),
+            ImmutableLinearStore::FileBacked(filebacked) => filebacked.stream_from(addr),
+            ImmutableLinearStore::Invalid => Err(std::io::ErrorKind::InvalidData.into()),
+        }
     }
 
-    fn size(&self) -> Result<u64, Error> {
-        self.deref().size()
+    pub fn size(&self) -> Result<u64, Error> {
+        match self {
+            ImmutableLinearStore::Historical(historical) => historical.size(),
+            ImmutableLinearStore::Proposed(proposed) => proposed.size(),
+            ImmutableLinearStore::FileBacked(filebacked) => filebacked.size(),
+            ImmutableLinearStore::Invalid => Err(std::io::ErrorKind::InvalidData.into()),
+        }
     }
 }
 
-/// Some linear stores support updates
-pub(super) trait WriteLinearStore: Debug {
-    fn write(&mut self, offset: u64, object: &[u8]) -> Result<usize, Error>;
+#[derive(Debug)]
+pub(super) enum MutableLinearStore {
+    Proposed(Proposed<Mutable>),
+    Invalid, // TODO do we need this variant? If not we can remove MutableLinearStore
```

Comment:
Yet another reason to remove Invalid. Let's make these invalids actually work, even though it is probably a VM error to use them. They can't be committed ever of course, and we will drop references to them once we know they can't be committed